pip/qsharp/_qsharp.py

```diff
@@ -219,18 +219,129 @@ def get_interpreter() -> Interpreter:
+
+    def check_eq(
+        self, state: Union[Dict[int, complex], List[complex]], tolerance: float = 1e-10
+    ) -> bool:
+        """
+        Checks if the state dump is equal to the given state. This is not mathematical equality,
+        as the check ignores global phase.
+
+        :param state: The state to check against, provided either as a dictionary of state indices to complex amplitudes,
+            or as a list of real amplitudes.
+        :param tolerance: The tolerance for the check. Defaults to 1e-10.
+        """
+        phase = None
+        # Convert a dense list of real amplitudes to a dictionary of state indices to complex amplitudes
+        if isinstance(state, list):
+            state = {i: val for i, val in enumerate(state)}
+        # Filter out zero states from the state dump and the given state based on tolerance
+        state = {k: v for k, v in state.items() if abs(v) > tolerance}
+        inner_state = {k: v for k, v in self.__inner.items() if abs(v) > tolerance}
+        if len(state) != len(inner_state):
+            return False
+        for key in state:
+            if key not in inner_state:
+                return False
+            if phase is None:
+                # Calculate the phase based on the first state pair encountered.
+                # Every pair of states after this must have the same phase for the states to be equivalent.
+                phase = inner_state[key] / state[key]
+            elif abs(phase - inner_state[key] / state[key]) > tolerance:
+                # This pair of states does not have the same phase,
+                # within tolerance, so the equivalence check fails.
+                return False
+        return True
+
+    def as_dense_state(self) -> List[complex]:
+        """
+        Returns the state dump as a dense list of complex amplitudes. This will include zero amplitudes.
+        """
+        return [self.__inner.get(i, complex(0)) for i in range(2**self.qubit_count)]
+
+
+class ShotResult(TypedDict):
+    """
+    A single result of a shot.
+    """
+
+    events: List[Output]
+    result: Any
+    messages: List[str]
+    matrices: List[Output]
+    dumps: List[StateDump]
+
+
+def eval(
+    source: str,
+    *,
+    save_events: bool = False,
+) -> Any:
     """
     Evaluates Q# source code.
 
     Output is printed to console.
 
     :param source: The Q# source code to evaluate.
-    :returns value: The value returned by the last statement in the source code.
+    :param save_events: If true, all output will be saved and returned. If false, they will be printed.
+    :returns value: The value returned by the last statement in the source code or the saved output if `save_events` is true.
     :raises QSharpError: If there is an error evaluating the source code.
     """
     ipython_helper()
 
+    results: ShotResult = {
+        "events": [],
+        "result": None,
+        "messages": [],
+        "matrices": [],
+        "dumps": [],
+    }
```

Comment:
I like the overall solution.

1. I don't love how we have 3 arrays with no ordering between them, and then we have another array with an order but no types. Can't I (as the user) have it all? I want my event list to be ordered AND typed. 

2. I also feel like we're fighting having to use actual Python features here 😆  . The following just feel more "native" and take advantage of Python's own dynamic typing rather than the dynamically-exposing-Rust-enums we're doing here.


```python
output for output in result.events if isinstance(output, Matrix)
```

```python
for output in result.events:
  if isinstance(output, Matrix):
    print(str(output))
```

Now granted that's a lot more code to write. OK, so let's forgo using actual Python classes. How about something like the below? We do already have `output.state_dump()`, so adding `matrix()` and `message()` feel like a logical next step.


```python
output for output in result.events if output .matrix() is not None
```

```python
for output in result.events:
  if output.matrix() is not None:
    print(str(output))
```

And if that's too wordy, maybe throw in a method for filtering on the fly

```python
result.events.matrices()   # A convenience method that filters on the events list, not a separately stored list
```

Reply:
Hmm... if `ShotResult` were a class I could definitely make this all work more nicely. I guess the reason I was gravitating toward keeping them in separate lists is that if I have code like this:
```qsharp
{
    Message("Foo");
    DumpMachine();
    Message("Bar");
    Message("Baz");
    DumpMachine();
}
```
I'm far more likely to think to myself, "I want the second dump," than, "I want the fifth output." Yes, I could do it with filtering and sublcassing and all that, but that requires that I know the right subclasses and that I write the filter, as you mention. So I'd want to put this all into a member function, but I can't do that on `TypedDict`. At least with this approach, if someone has their `results` and they try to use `results[0][...]` with autocomplete on the "..." they'll get to see the options of whats in there without having to know the different ways they could filter the "events" entry that has anything.

Frustratingly, the real solution here is a class... which is a breaking change :(

Reply:
I think compatibility with `qsharp_widgets` limits our options quite a bit here, so I'm happy with whatever you can come up with given the constraints.

Reply:
I ended up landing at a middle ground of keeping the type specific lists while also making the overall "events" list have type converted entries.